workorder-job-repeater.lua

```diff
@@ -0,0 +1,219 @@
+        if posting.job == nil then
+            -- It's free real estate!
+            posting.job = job
+            posting.anon_1 = 0
+            posting.flags.dead = false
+            
+            job.posting_index = posting.idx
+            addedIndex = posting.idx
+            break
+        end
+    end
+    if addedIndex then
+        log(DEBUG, "Added job " .. job.id .. " to postings as index " .. addedIndex)
+    else
+        log(DEBUG, "Couldn't find posting slot to add in job " .. job.id)
+    end
+    
+    return addedIndex
+end
+
+local function repeatJob(order, job)
+    if not (job and order) then
+        qerror('! (job and order)')
+        return
+    end
+    
+    if only_infty and 0 ~= order.amount_total then
+        log(DEBUG, "skipping job " .. dfhack.job.getName(job) .. " (id " .. job.id .. "): order.amount_total > 0")
+        return
+    end
+    
+    if not (order.status.validated and order.status.active) then
+        log(DEBUG, "skipping job " .. dfhack.job.getName(job) .. " (id " .. job.id .. ") from order " .. order.id .. " because of order.status")
+        return
+    end
+    
+    -- for debug purposes: pause the game
+    --df.global.pause_state = true
+
+    log(DEBUG, "Repeat job " .. dfhack.job.getName(job) .. " (id " .. job.id .. ") from order " .. order.id)
+    
+    -- we need a clone because original(which itself is a copy) is destroyed or something.
+    -- cloneJobStruct doesn't copy flags over.
+    local by_manager = job.flags.by_manager
+    local do_now = job.flags.do_now
+    job = dfhack.job.cloneJobStruct(job)
+    job.flags.by_manager = by_manager
+    job.flags.do_now = do_now
+    
+    -- step 0: adjust what we can right now
+    job.flags.working = false
+    job.completion_timer = -1
+    job.items:resize(0) -- remove old items we worked upon
+    
+    local workshop = nil
+    local ixs = {}
+    for ix, gref in ipairs(job.general_refs) do
+        if gref:getType() == df.general_ref_type.BUILDING_HOLDER then
+            local building = gref:getBuilding()
+            if building then
+                log(DEBUG, "Building from gref: " .. building.id)
+                workshop = building
+            else
+                -- should never ever happen, I think
+                log(WARN, "Building from gref not found: ", gref.building_id)
+            end
+        else
+            ixs[ix] = true
+        end
+    end
+    for ix in pairs(ixs) do
+        log(DEBUG, "Removing from gref: ", job.general_refs[ix])
+        job.general_refs:erase(ix)
+    end
+    log(DEBUG, "#gref: " .. #job.general_refs .. " (should be 1)")
+    
+    -- step 1: associate a posting with it
+    local ok = addJobToPostings(job)
+    if not ok then
+        log(WARN, "Couldn't add job of order.id " .. order.id .. " to a posting")
+        return
+    end
+    
+    -- step 2: add the job to the job list
+    local ok = dfhack.job.linkIntoWorld(job, false) -- we don't need a new id, I think?
```

Comment:
The main reason is to reuse it is because we can. Also, this mimics a job on repeat more closely -- which may or may not influence how the game treats these jobs. The id isn't bound to the original job struct and just stays the same.

Since it's working fine (as far as I can tell at least) I'd say we leave it as is.

Reply:
👍